src/Build/Logging/BinaryLogger/BinaryLogger.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.IO;
+using System.IO.Compression;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// A logger that serializes all incoming BuildEventArgs in a compressed binary file (*.binlog). The file
+    /// can later be played back and piped into other loggers (file, console, etc) to reconstruct the log contents
+    /// as if a real build was happening. Additionally, this format can be read by tools for
+    /// analysis or visualization. Since the file format preserves structure, tools don't have to parse
+    /// text logs that erase a lot of useful information.
+    /// </summary>
+    public class BinaryLogger : ILogger
+    {
+        private const int FileFormatVersion = 1;
+
+        private Stream stream;
+        private BinaryWriter binaryWriter;
+        private BuildEventArgsWriter eventArgsWriter;
+
+        private string FilePath { get; set; }
+
+        /// <summary>
+        /// The binary logger Verbosity is always maximum (Diagnostic). It tries to capture as much
+        /// information as possible.
+        /// </summary>
+        public LoggerVerbosity Verbosity { get; set; } = LoggerVerbosity.Diagnostic;
+
+        /// <summary>
+        /// The only supported parameter is the output log file path (e.g. "msbuild.binlog") 
+        /// </summary>
+        public string Parameters { get; set; }
+
+        /// <summary>
+        /// Initializes the logger by subscribing to events of IEventSource
+        /// </summary>
+        public void Initialize(IEventSource eventSource)
+        {
+            Environment.SetEnvironmentVariable("MSBUILDTARGETOUTPUTLOGGING", "true");
+
+            ProcessParameters();
+
+            try
+            {
+                stream = new FileStream(FilePath, FileMode.Create);
+            }
+            catch (Exception e)
+            {
+                string errorCode;
+                string helpKeyword;
+                string message = ResourceUtilities.FormatResourceString(out errorCode, out helpKeyword, "InvalidFileLoggerFile", FilePath, e.Message);
+                throw new LoggerException(message, e, errorCode, helpKeyword);
+            }
+
+            stream = new GZipStream(stream, CompressionLevel.Optimal);
+            binaryWriter = new BinaryWriter(stream);
+            eventArgsWriter = new BuildEventArgsWriter(binaryWriter);
+
+            binaryWriter.Write((byte)FileFormatVersion);
```

Comment:
I would store your version as a byte in the first place, otherwise I could set it to something too big and it would break this logger right?